python/lsst/cp/pipe/ptc/cpPtcSolve.py

```diff
@@ -1077,14 +1111,153 @@ def _makeZeroSafe(self, array, substituteValue=1e-9):
 
         return array
 
+    def fitPtcRolloff(self, dataset):
+        """Fit the photon transfer curve to the
+        Astier+19 approximation (Eq. 16) with a roll-off model to try
+        and capture saturation effects.
+
+        Parameters
+        ----------
+        dataset : `lsst.ip.isr.ptcDataset.PhotonTransferCurveDataset`
+            The dataset containing the means, variances and
+            exposure times.
+
+        Returns
+        -------
+        dataset : `lsst.ip.isr.ptcDataset.PhotonTransferCurveDataset`
+            This is the same dataset as the input parameter, however,
+            it has been modified to include information such as the
+            fit vectors and the fit parameters. See the class
+            `PhotonTransferCurveDataset`.
+        """
+        # For FULLCOVARIANCE model fit
+        matrixSideFit = dataset.covMatrixSideFullCovFit
+        nanMatrixFit = np.empty((matrixSideFit, matrixSideFit))
+        nanMatrixFit[:] = np.nan
+
+        def errFunc(p, x, y):
+            return ptcFunc(p, x) - y
+
+        for i, ampName in enumerate(dataset.ampNames):
+            meanVec = dataset.rawMeans[ampName].copy()
+            varVec = dataset.rawVars[ampName].copy()
+            varVec = self._makeZeroSafe(varVec)
+
+            # Make sure we have this properly sorted.
+            goodPoints = dataset.expIdMask[ampName].copy()
+
+            if not goodPoints.any():
+                msg = (f"SERIOUS: All points in goodPoints: {goodPoints} "
+                       f"are bad. Setting {ampName} to BAD.")
+                self.log.warning(msg)
```

Comment:
Is this a duplicate warning?  E.g., if we did the turnoff check before and it gave this warning it's not fun to see it doubled for all bad amps.  (Same for the one just below).  E.g., aren't these already filled bad, etc?